### PR TITLE
Speed up (up to 7x!) building large system libraries

### PR DESCRIPTION
**Pull request by RReverser**
Large system libraries like libc consist of lots of individual objects.

Previously, we'd compile each of them individually, although the processes themselves were parallelised across CPU cores.

However, this resulted in a lot of per-object overhead: create a Python process for Emscripten itself, compile our Python code, execute a bunch of custom logic in said Python code, create another process for Clang with the computed flags, and finally get an object. This overhead is even more pronounced on Windows where (leaving aside details on why) creating a new process is much more expensive than on Linux.

Luckily, Clang, and, by extension, Emscripten, allows to compile lots of source files into object files in the same command invocation / same process, as long as all of them share same compilation flags and target directory (which has to be current working directory).

In this PR I'm leveraging that feature by grouping all the sources within a library by compiler and flags they use. Then, I'm splitting said groups into chunks. Chunk size is currently chosen as `2 * number_of_cores` as a heuristic to strike balance between creating large batches and avoiding workload imbalance where some CPU core might've already finished and stays idle while others are still processing their share of source files. Finally, each chunk is added to the list of commands, and all of those commands are executed in parallel across CPU cores like before. (There is still one exception for commands that need to be singled out, but I won't go into those details here.)

On my laptop with 8 cores, this leads to the following speedups on large system libraries when building `./em++ temp.cpp -s STRICT` where `temp.cpp` is just `int main() {}` and the cache is empty:

| Library Name      | Before (Windows) | After (Windows) | Before (Linux) | After (Linux) | Speedup (Windows) | Speedup (Linux) |
|-------------------|------------------|-----------------|----------------|---------------|-------------------|-----------------|
| libc-debug.a      | 141.7s           | 29.6s           | 53.2s          | 7.7s          | 4.8x              | 7.0x            |
| libcompiler_rt.a  | 22.7s            | 6.6s            | 8.7s           | 1.8s          | 3.5x              | 4.8x            |
| libc++-noexcept.a | 39.5s            | 37.0s           | 20.8s          | 18.6s         | 1.1x              | 1.1x            |

### CODE REVIEW

**Comment by RReverser** [comment]
I'm not seeing those Windows CI failures on Windows locally but I suspect I know why it's happening (seen the same in recent Ninja PR), should be easy to fix.

File: tools/system_libs.py
Comment (kripken) [inline tools/system_libs.py:515]:
```suggestion
        # If we have seen a file with the same name before, we are on a case-insensitive
```

**Review by kripken — APPROVED** [review]
Wow, excellent idea! And great speedups!

(I didn't even know you can pass in multiple filenames with `-c`...)

**Comment by RReverser** [comment]
> (I didn't even know you can pass in multiple filenames with `-c`...)

It's actually something I only learnt a few days ago myself, and immediately was like "wait a second... I know where this would be useful" 😅

**Comment by sbc100** [comment]
Nice!   

I guess its a good job that clang doesn't using parallelism internally compile the different source files that its passed, otherwise we would end up with N * N.

**Comment by sbc100** [comment]
Note that when you have multiple inputs in a single compile command you loose the ability to specify output filenames.   So things like `clang -c dir1/foo.c and dir2/foo.c` won't work since both sources compile to `foo.o`.

Luckily I guess in this case we don't care much about the names of the object files, and we address name collisions as they arise.

**Comment by RReverser** [comment]
> I guess its a good job that clang doesn't using parallelism internally compile the different source files that its passed, otherwise we would end up with N * N.

I don't think they would do that implicitly, but I actually searched for it and saw a PR where someone was adding parallelism for clang-cl and said they could do it for clang itself via `-j` if there was interest, but that didn't go anywhere.

It would be nice if Clang has such opt-in parallelism since then we wouldn't even need to spawn bunch of processes and could instead pass all the source files in a batch to a single Clang command and let it build those in parallel threads (which are much more lightweight), making builds even faster... Oh well.

**Comment by RReverser** [comment]
> Note that when you have multiple inputs in a single compile command you loose the ability to specify output filenames. So things like `clang -c dir1/foo.c and dir2/foo.c` won't work since both sources compile to `foo.o`.

Yup, I know, that's why I still have that exemption in code where commands that would result in same object file are split out and handled separately.

File: emcc.py
Comment (sbc100) [inline emcc.py:3632]: Do you use this flag?

Comment (RReverser) [reply emcc.py:3632]: No, but couple of the tests do, so invoking subprocess with a different working directory broke them.

**Comment by RReverser** [comment]
Had to fix one more issue this uncovered (see comment in code), and I think this is good to go.

File: emcc.py
Comment (RReverser) [reply emcc.py:3632]: Although I now wonder if environment variable `EM_CACHE` (and maybe some others like `EM_CONFIG`, `EMCC_TEMP_DIR` etc?) might be subject to the same issue, just untested? Do people set those variables to relative paths / is it officially supported?

File: tools/system_libs.py
Comment (sbc100) [inline tools/system_libs.py:541]: Is there a reason to copy cmd here?

File: tools/system_libs.py
Comment (sbc100) [inline tools/system_libs.py:91]: This functions is also used to build ports (see `ports/__init__.py`) so maybe keep the logging in this function add add an options "num_inputs" argument?

**Comment by sbc100** [comment]
I'm surprised the speedup is bigger on linux, given that windows is that one that we expect to see have a higher cost of process creation.

File: tools/system_libs.py
Comment (sbc100) [reply tools/system_libs.py:541]: Ah, I didn't notice the conversion to tuple above

**Comment by RReverser** [comment]
> I'm surprised the speedup is bigger on linux, given that windows is that one that we expect to see have a higher cost of process creation.

I think this is one of the cases where it should be looked at not so much a multiplier kind of increase (new time / previous time like I did), as absolute reduction of `overhead * number_of_files.`.

So e.g. `libc-debug` which has 1024 files reduced overhead by (141.7-29.6)/1024=0.1 seconds per file on Windows and (53.2-7.7)/1024=0.04 seconds on Linux.

Looking at it this way scales more correctly when we look at other libraries - e.g. for `libcompiler_rt` which has 175 files we get (22.7-6.6)/175=0.09 seconds reduction on Windows and (8.7-1.8)/175=0.04 seconds on Linux, basically same time per file.

File: tools/system_libs.py
Comment (RReverser) [reply tools/system_libs.py:91]: Sgtm.

**Comment by dschuff** [comment]
One thing to watch out for on testing here: the CircleCI builder actually uses ninja to prebuild the libraries, so IIUC that's not covering this change. Can you maybe try a commit that removes EMCC_USE_NINJA from the .config.yml and tests that way? It may still be slower than ninja so we may want to revert it, but at least we can test it before landing. The emscripten-releases CI uses this builtin way, so we will get coverage, but not until after it lands here.

**Comment by sbc100** [comment]
If this method ends up being faster than ninja I wonder if we should just remove the EMCC_USE_NINJA support?

**Comment by RReverser** [comment]
> One thing to watch out for on testing here: the CircleCI builder actually uses ninja to prebuild the libraries, so IIUC that's not covering this change. Can you maybe try a commit that removes EMCC_USE_NINJA from the .config.yml and tests that way? It may still be slower than ninja so we may want to revert it, but at least we can test it before landing. The emscripten-releases CI uses this builtin way, so we will get coverage, but not until after it lands here.

It's still much faster than Ninja, yes.

I actually initially implemented this change for Ninja as I'm using it locally as well. But I got there only partially because Ninja doesn't have built-in support for batching yet (https://github.com/ninja-build/ninja/issues/1004) and the workaround to find out which files have changed so that only those are batched is rather involved - you need to have task that just collects changed filenames into a temporary file, then chunk that temporary files into batches... it gets complicated.

So I decided to stop there and implement for non-Ninja version of the building script instead, which is both much easier implementation-wise and benefits more users as most didn't opt-in to Ninja yet.

I'll try to remove EMCC_USE_NINJA from config.

File: emcc.py
Comment (RReverser) [reply emcc.py:3632]: Thoughts on this one - should I worry about other relative paths?

**Comment by RReverser** [comment]
> One thing to watch out for on testing here: the CircleCI builder actually uses ninja to prebuild the libraries, so IIUC that's not covering this change.

If this is the case (which it seems to be), I'm a bit surprised some tests failed before due to a change in this non-Ninja path (due to me spawning subprocesses in a different working directory when building system library)...

File: emcc.py
Comment (sbc100) [reply emcc.py:3647]: Its hard to imagine folks using relative paths in EM_CACHE and EM_TEMP_DIR environment variables.   

I wouldn't worry about that case.

Comment (RReverser) [reply emcc.py:3647]: Okay, fingers crossed.

**Comment by dschuff** [comment]
> > One thing to watch out for on testing here: the CircleCI builder actually uses ninja to prebuild the libraries, so IIUC that's not covering this change.
> 
> If this is the case (which it seems to be), I'm a bit surprised some tests failed before due to a change in this non-Ninja path (due to me spawning subprocesses in a different working directory when building system library)...

It only uses ninja for the 'build libraries' step, before tests run But that doesn't build every configuration of every library. IIRC there are still some libraries that get built on-demand during testing runs, and we don't use ninja for that.

**Comment by RReverser** [comment]
@dschuff Looks like it passes w/o EMCC_USE_NINJA too. CI seems to have a lot of build time variation on commit-to-commit basis though, so it's hard to judge if it's any faster there, but at least it still works.

**Comment by RReverser** [comment]
One more thing I could probably do is share this logic for ports as well, as I see some ports can also be pretty slow (e.g. libsqlite3 takes 47.57s and libicu takes 38.98s).

But I'd rather leave this for a separate follow-up PR.

**Comment by dschuff** [comment]
OK great. I think it probably makes sense to check this in without changing the test configuration for now, and we can experiment with that in a separate PR. Assuming Sam is OK with the actual change, I don't even mind pushing it in before the tests finish after you revert the config bit :)

**Comment by dschuff** [comment]
And yeah I agree that speeding up ports would be great in a followup PR, and that's something I didn't get around to with ninja.

**Comment by RReverser** [comment]
Ok I'm going to address Sam's comment above first and revert Ninja change. I think I'll still wait for Sam's review / approval before enabling auto-merge though.

**Comment by RReverser** [comment]
> And yeah I agree that speeding up ports would be great in a followup PR, and that's something I didn't get around to with ninja.

Out of curiosity, why do ports even have a different codepath and implementation of building instead of reusing the `Library` class? Are they conceptually different from system libraries?

**Comment by sbc100** [comment]
> > And yeah I agree that speeding up ports would be great in a followup PR, and that's something I didn't get around to with ninja.
> 
> Out of curiosity, why do ports even have a different codepath and implementation of building instead of reusing the `Library` class? Are they conceptually different from system libraries?

Mostly just the history of how they evolved I think.

One big difference is that ports are one-per-file (the idea was to allow folks to add new ones by just creating new files in tools/ports.   Whereas system libraries are all declared in the same time.

**Review by sbc100 — APPROVED** [review]

**Comment by RReverser** [comment]
> One big difference is that ports are one-per-file (the idea was to allow folks to add new ones by just creating new files in tools/ports.

Okay, but the build process for objects + archive is roughly the same?

**Comment by kripken** [comment]
@RReverser , thank you so much for this change, I can hardly notice building `libc` locally any more, it's so fast! :smile:

**Comment by RReverser** [comment]
Thanks! I was surprised to see how fast it worked too, I had to triple-check it's still producing a working library with all the same objects several times before making a PR 😅

**Comment by RReverser** [comment]
I wonder if I should've added this to changelog... OTOH I guess it's not really a functional change.

**Comment by dschuff** [comment]
Nice! On the emscripten-releases builders, this change cut full-library build time almost in half on mac (~41m to ~22m), by 35% on Windows (28m to 18m) and by 32% on Linux (19m to 13m).

**Comment by RReverser** [comment]
> I think the main thing preventing the really big speedups for these cases is that embuilder builds each library sequentially, so for the large number of small libraries, each one is still limited by the time it takes to build the slowest/largest file. That's the problem that the ninja build fixes, but this solution is still great for on-demand library builds, which Ninja doesn't support.

I was actually playing with this a bit more and found couple more opportunities for improvements. Nothing as dramatic as here, but I can get squeeze ~15% more for regular build, and I can get Ninja in the vicinity of those timings above as well even w/o batching. I'll send some PRs later.

As for compilation of tiny libraries, I thought about doing that (but didn't experiment with that in particular, so if you want to give it a go, please do), but it will break individual timings and logs per library - not sure how important those are?

**Comment by RReverser** [comment]
That said, I'm surprised Ninja (the current implementation at least) is faster than this approach anywhere. It was much worse on my 8-core machine, but I guess with large number of cores it might scale worse...

**Comment by RReverser** [comment]
Ah sorry, re-reading, you're saying that Ninja is slower when you meaure all libraries built together, not any individual library? Yeah that makes sense.

**Comment by RReverser** [comment]
> I think the main thing preventing the really big speedups for these cases is that embuilder builds each library sequentially, so for the large number of small libraries, each one is still limited by the time it takes to build the slowest/largest file. That's the problem that the ninja build fixes

Ok I'm probably misunderstanding something. When I use EMCC_USE_NINJA, it seems that embuilder still builds individual libraries one by one. So how does ninja build fix this?

**Comment by curiousdannii** [comment]
Sounds like this will negate a lot of the impact of https://github.com/emscripten-core/emsdk/issues/660? Which will be awesome, thanks!

**Comment by RReverser** [comment]
I mean, fixing caching would be still useful, but yeah, it should help :)

> When I use EMCC_USE_NINJA, it seems that embuilder still builds individual libraries one by one. So how does ninja build fix this?

@dschuff I think I understant what you meant. There are couple of problems there:

1. Those `./embuilder` "groups" are not documented in `--help` (other than `ALL`), I didn't even know they existed before looking into the source code.
2. Regular `./emcc` / `./em++` doesn't use those groups and builds libraries one by one, which means most users don't benefit from this optimisation even if they opted into Ninja.
3. The logs from such grouped `./embuilder` are somewhat misleading because it still shows individual steps as if it builds separate libraries, but those steps take ~0 seconds as it actually only generates Ninja file.

**Comment by RReverser** [comment]
Did some more measurements and that's still odd that the difference so big in the other direction.

For example, on my 8-core machine `./embuilder build MINIMAL` takes 11:55 min with Ninja and 06:36 min without Ninja. I see in each case ~2 min are spent on sqlite3 ports (which are executed in blocking way for now in each build system), so the difference for parts that actually change (system libs) is more like 10 min vs 4.5 min, not in favour of Ninja.

Does it scale that much worse with 16 cores that Ninja becomes faster instead?

**Comment by sbc100** [comment]
Ah, I forgot one more reason why I'd like to keep/improve `EMCC_USE_NINJA`: incremental builds.  I been working today on a single file in libc with the lot of iteration.   I'd really like to to be able to have ninja just rebuild a single file in libc on each iteration.   This means keeping the object tree around which I'm not sure we support yet, but it would have saved me a lot of time today.

**Comment by RReverser** [comment]
> Ah, I forgot one more reason why I'd like to keep/improve `EMCC_USE_NINJA`: incremental builds.

Yup, that's definitely something only Ninja can do currently.

**Comment by RReverser** [comment]
> I'd really like to to be able to have ninja just rebuild a single file in libc on each iteration. This means keeping the object tree around which I'm not sure we support yet, but it would have saved me a lot of time today.

Hm I thought it did? I think I remember being able to cd into the dir and do `ninja` manually, but maybe not?

**Comment by sbc100** [comment]
> > I'd really like to to be able to have ninja just rebuild a single file in libc on each iteration. This means keeping the object tree around which I'm not sure we support yet, but it would have saved me a lot of time today.
> 
> Hm I thought it did? I think I remember being able to cd into the dir and do `ninja` manually, but maybe not?

Maybe that method works.   I only tried `rm /path/to/libc.a && make` which always did a clean build.

**Comment by sbc100** [comment]
One downside of this is my error messages not contain filenames that are not relative the emscripten directory:

```
../../../system/lib/gl/webgl1.c:57:54: error: incompatible integer to pointer conversion passing 'EMSCRIPTEN_WEBGL_CONTEXT_HANDLE' (aka 'long') to parameter of type 'const void *' [-Wint-conversion]
```

Rather than before:

```
system/lib/gl/webgl1.c:57:54: error: incompatible integer to pointer conversion passing 'EMSCRIPTEN_WEBGL_CONTEXT_HANDLE' (aka 'long') to parameter of type 'const void *' [-Wint-conversion]
```

Not a huge deal but means it a little harder to just copy and paste those filenames.

**Comment by RReverser** [comment]
It would be possible to keep the full absolute paths, but it would mean using potentially more expensive response file more often.

Given that most users don't need this and expect system libs to be prebuilt, idk if it's worth it vs suggesting that contributors use Ninja, for example.

**Comment by RReverser** [comment]
Ah actually you could also just pass `-fdiagnostics-absolute-paths` either via EMCC_CFLAGS or even add it to the build scripts.

**Comment by sbc100** [comment]
> Ah actually you could also just pass `-fdiagnostics-absolute-paths` either via EMCC_CFLAGS or even add it to the build scripts.

What I wish we could do is avoid `chdir`ing into the build directory and instead do something like `-o /dir/for/object/` when building multiple `.c` files, but sadly I don't think clang supports that.

**Comment by RReverser** [comment]
Yeah Clang supports `-working-directory=...` (which emcc doesn't) which would allow changing cwd from the Python script, but wouldn't solve your issue.

I think `-fdiagnostics-absolute-paths` is your best bet as you at least paths that can be easily clicked on / opened in a terminal regardless of cwd.

**Comment by sbc100** [comment]
I just had a thought that we could also speed this up by not spawning N clang process from emcc.py when `-c` is used with mutiple file names.  Sadly we will likely still need to do then when not using `-c`.. but it might be worth having two code paths.

**Comment by RReverser** [comment]
> I just had a thought that we could also speed this up by not spawning N clang process from emcc.py when `-c` is used with mutiple file names. Sadly we will likely still need to do then when not using `-c`.. but it might be worth having two code paths.

I actually have a branch with that too 😅 Although that didn't yield nearly as noticeable speedup, which suggests it's primarily Python execution that was taking majority of saved overhead time and not as much process creation as initially thought. Doesn't harm to send it anyway I guess.

But I wonder if a better long-term solution could be to somehow cache cflags in LRU cache inside EMSDK_CACHE (is it even possible to do reliably or is it too dependant on working dir, environment and such?). This way not only system libraries, but any large project using Emscripten could benefit from the same saved time of executing bunch of Python code just to calculate cflags for all the individual `-c` invocations.

**Comment by RReverser** [comment]
> Yes, somekind of python-overhead profiling would probably in order here.

Yeah I did proper profiling too after this PR and found few more minor opportunities like described above:

> I was actually playing with this a bit more and found couple more opportunities for improvements. Nothing as dramatic as here, but I can get squeeze ~15% more for regular build, and I can get Ninja in the vicinity of those timings above as well even w/o batching. I'll send some PRs later.

Which had to do with process scheduling, but the rest looks a bit too evenly distributed to single out a single cause which is why caching similar to `emcc --cflags` seems like the most sensible suggestion for now. It would even help all those build systems that pass explicit `-o` for object files.

**Comment by RReverser** [comment]
> Yes, but remember that `--cflags` depends on bunch of `-s` settings to, so would it would to include all of those in the cache key.

Yeah I did imagine that it would be `(emcc cflags) -> (clang cflags)` cache + whatever other necessary variables added to the key. So this would speed up building any large libraries where cflags and working dir are already repetitive and shared between object files, regardless of the build system used.

**Comment by sbc100** [comment]
> > Are you imagining that a lot of the python work that emcc does is simply to calculate those flags? I would have though most of the overhead is purly in starting the compiler and loading all of the `.py` file. Are you imagining that we could skip that somehow if we have the cflags cached?
> 
> All good questions... I would expect that we'd be able to start with just one `.py` that looks up stuff in cache and loads & executes the rest of the code only if it's not there, so it would avoid all of the mentioned overheads combined. But maybe it's really just starting Python that is so slow? That I'm not sure about.

Yup, for that it would probably make sense to profile.

**Comment by RReverser** [comment]
> What part of the build cost are you proposing to save by using this cache? emcc parsing its command line flags?

emcc does a lot more than just parse them though, before getting to the clang stage.

**Comment by sbc100** [comment]
> > What part of the build cost are you proposing to save by using this cache? emcc parsing its command line flags?
> 
> emcc does a lot more than just parse them though, before getting to the clang stage.

Yes, and we can look at reducing that if it has measurable impact.

**Comment by sbc100** [comment]
> > > What part of the build cost are you proposing to save by using this cache? emcc parsing its command line flags?
> > 
> > 
> > emcc does a lot more than just parse them though, before getting to the clang stage.
> 
> Yes, and we can look at reducing that if it has measurable impact.

My assumption has always been the this wasn't significant compared to the cost of actually running clang.. but I could well be wrong about that.

**Comment by RReverser** [comment]
Well I did mean above is that I profiled it + experimented with just generating Ninja file that invokes Clang directly with pre-computed cflags, and that resulting in almost the same improvement (for Ninja) as this PR did for non-Ninja build, with very little overhead left (what's left can be attributed to Clang process creation x number of files).

Which strongly suggests that most of the overhead reduced here did come from executing the Python code, and actually invoking Clang is relatively cheap.

**Comment by RReverser** [comment]
> Which strongly suggests that most of the overhead reduced here did come from executing the Python code, and actually invoking Clang is relatively cheap.

That also matches the observation that currently emcc.py will invoke separate clang.exe per each source even for `-c ...multiple files...`, so all this PR did is merged computation of cflags up to that point and it clearly yielded significant speedup.

That computation is the same overhead I think caching would help to reduce more universally.

**Comment by sbc100** [comment]
> > Which strongly suggests that most of the overhead reduced here did come from executing the Python code, and actually invoking Clang is relatively cheap.
> 
> That also matches the observation that currently emcc.py will invoke separate clang.exe per each source even for `-c ...multiple files...`, so all this PR did is merged computation of cflags up to that point and it clearly yielded significant speedup.
> 
> That computation is the same overhead I think caching would help to reduce more universally.

I imagine there two sources of overhead here:
(1) the cost of starting python.exe to run the very first line of python code in emcc.py
(2) the cost running the emcc.py python code up until subprocess creation

I would love to get rid of both of them one day, but you are right that getting rid of (2) could be done by cache the clang arguments.

We've also not spent much time to optimizing (2) so there could be a lot of other low hanging fruit here even without caching.

**Comment by RReverser** [comment]
> I imagine there two sources of overhead here:
(1) the cost of starting python.exe to run the very first line of python code in emcc.py
(2) the cost running the emcc.py python code up until subprocess creation

Fair enough, it's plausible. I'm not very familiar with Python execution and don't know what the cost of (1) looks like. I only measured (2) the actual python code.

> We've also not spent much time to optimizing (2) so there could be a lot of other low hanging fruit here even without caching.

Right, caching (as usual I guess heh) is more of a suggested band-aid. I certainly agree properly optimising (2) is better long-term for both cached and uncached scenarios, I just figured caching might be much quicker to implement to get them sooner.

**Pull request merged**
